CI: Build check with various optimization levels

ChinYikMing opened the pull request:
The submodules are built locally, which may lead to build failures. Specifically, the dtc submodule fails to build when using GCC versions 11.4.0 or 12.3.0 with optimization levels set to -O0 or with debug flag -g. This is a critical issue, as it can disable the debug mode. To address this, optimization levels and debug level build are introduced to stabilize the pull request changes.

The compiled error:
/usr/bin/ld: /tmp/cc2eaHdM.ltrans0.ltrans.o: in function fdt_del_node': <artificial>:(.text+0x279f0): undefined reference to fdt_node_end_offset_' collect2: error: ld returned 1 exit status

Note: I have only used gcc 11.4.0 and gcc 12.3.0 to test this out. Github runners should test other compilers.

After bisecting, the original cause was found after #534.

ChinYikMing committed:
CI: Build check with various optimization levels

The submodules are built locally, which may lead to build failures.
Specifically, the dtc submodule fails to build when using GCC versions
11.4.0 or 12.3.0 with optimization levels set to -O0 or with debug flag
-g. This is a critical issue, as it can disable the debug mode. To
address this, optimization levels and debug level build are introduced
to stabilize the pull request changes.

The compiled error:
/usr/bin/ld: /tmp/cc2eaHdM.ltrans0.ltrans.o: in function fdt_del_node':
<artificial>:(.text+0x279f0): undefined reference to fdt_node_end_offset_'
collect2: error: ld returned 1 exit status

Note: I have only used gcc 11.4.0 and gcc 12.3.0 to test this out.
Github runners should test other compilers.

After bisecting, the original cause was found after #534.

ChinYikMing commented:
See [action](https://github.com/sysprog21/rv32emu/actions/runs/12965405707/job/36165100731) for more information.

Clang has no problem with various optimization build. Interesting.

RinHizakura commented:
This is not the right way to fix it. The issue could happen because I only compile the object that I think we need, but maybe more objects are required for `-g` + `-O0`. However, this can't explain the different results on Clang and GCC that you mentioned. I also have no idea here.
 
Can you check if you can also fix this with the following change?

```diff
--- a/mk/system.mk
+++ b/mk/system.mk
@@ -28,7 +28,7 @@ DEV_OBJS := $(patsubst $(DEV_SRC)/%.c, $(DEV_OUT)/%.o, $(wildcard $(DEV_SRC)/*.c
 deps := $(DEV_OBJS:%.o=%.o.d)

 OBJS_EXT += system.o
-OBJS_EXT += dtc/libfdt/fdt.o dtc/libfdt/fdt_ro.o dtc/libfdt/fdt_rw.o
+OBJS_EXT += dtc/libfdt/fdt.o dtc/libfdt/fdt_ro.o dtc/libfdt/fdt_rw.o dtc/libfdt/fdt_wip.o
```

RinHizakura commented:
Oh sorry, not aware that you are just trying to robust the CI/CD here instead of fixing anything. We can accept the test first and fix it in another PR.

RinHizakura commented on .github/workflows/main.yml:
```diff
@@ -58,16 +58,107 @@ jobs:
+        CC: ${{ steps.install_cc.outputs.cc }}
+      run: |
+            make distclean
+            make OPT_LEVEL=-Ofast ENABLE_SYSTEM=1 -j$(nproc)
```
Not sure if we need to take care of `-Os` or even just retain the presentative -O level only, otherwise the possible options may be too many. Will let @jserv decide this.

ChinYikMing replied:
One more `-Oz` can be the candidate. It is related to size just like `-Os`. Thus, not embracing them at first.

ChinYikMing commented:
> Oh sorry, not aware that you are just trying to robust the CI/CD here instead of fixing anything. We can accept the test first and fix it in another PR.

Agree. Thanks! Since with default `-O2` build, the problem does not exist. I revealed this problem when trying to do some debugging.

ChinYikMing commented:
> This is not the right way to fix it. The issue could happen because I only compile the object that I think we need, but maybe more objects are required for `-g` + `-O0`. However, this can't explain the different results on Clang and GCC that you mentioned. I also have no idea here.
> 
> Can you check if you can also fix this with the following change?
> 
> ```diff
> --- a/mk/system.mk
> +++ b/mk/system.mk
> @@ -28,7 +28,7 @@ DEV_OBJS := $(patsubst $(DEV_SRC)/%.c, $(DEV_OUT)/%.o, $(wildcard $(DEV_SRC)/*.c
>  deps := $(DEV_OBJS:%.o=%.o.d)
> 
>  OBJS_EXT += system.o
> -OBJS_EXT += dtc/libfdt/fdt.o dtc/libfdt/fdt_ro.o dtc/libfdt/fdt_rw.o
> +OBJS_EXT += dtc/libfdt/fdt.o dtc/libfdt/fdt_ro.o dtc/libfdt/fdt_rw.o dtc/libfdt/fdt_wip.o
> ```

Tested it out, fixed it! Another commit will be push later.

ChinYikMing commented:
@RinHizakura  One more thing, I noticed the `dtc` submodule has integrated the `dtc` compiler, shall we use it? So that, the new user does not need to install the `dtc` on their own.

ChinYikMing committed:
Fix dtc submodule build error when using gcc -g or -O0

The build error:
/usr/bin/ld: /tmp/cc2eaHdM.ltrans0.ltrans.o: in function fdt_del_node':
<artificial>:(.text+0x279f0): undefined reference to fdt_node_end_offset_'
collect2: error: ld returned 1 exit status

ChinYikMing commented:
The action that passed the various optimization level build after apply commit 139af81: [link](https://github.com/sysprog21/rv32emu/actions/runs/12965695115).

otteryc commented:
Link-Time Optimization (LTO) may be the culprit when building with `-O0` flag. GCC worked fine after adding the `-fno-lto` flag.

ChinYikMing commented:
> Link-Time Optimization (LTO) may be the culprit when building with `-O0` flag. GCC worked fine after adding the `-fno-lto` flag.

Without modifying the lto configuration, adding this object `dtc/libfdt/fdt_wip.o` solves it.

RinHizakura commented:
> One more thing, I noticed the dtc submodule has integrated the dtc compiler, shall we use it? So that, the new user does not need to install the dtc on their own.

I think we don't need to introduce such complications because building dtc may require other dependencies.

jserv reviewed (changes requested):
Rebase the latest `master` branch to resolve Arm64 host breakage on CI.

jserv reviewed (commented):
<!-- github-benchmark-action-comment(start): Benchmarks Summary -->
# Benchmarks

<details>

| Benchmark suite | Current: 1a2bcc9cd1b1c18c5d9b7d46a74fba63b075cf0a | Previous: 64ea138c99ecfa1a91a4a26e23226867ac857cb5 | Ratio |
|-|-|-|-|
| `Dhrystone` | `1313` Average DMIPS over 10 runs | `1324` Average DMIPS over 10 runs | `1.01` |
| `Coremark` | `962.001` Average iterations/sec over 10 runs | `969.943` Average iterations/sec over 10 runs | `1.01` |

</details>

This comment was automatically generated by [workflow](https://github.com/sysprog21/rv32emu/actions?query=workflow%3ABenchmark) using [github-action-benchmark](https://github.com/marketplace/actions/continuous-benchmark).
<!-- github-benchmark-action-comment(end): Benchmarks Summary -->

jserv commented:
I am concerned that the proposed changes increase the CI pipeline completion time. Could we rework this for two types of builds: one without optimizations (using `-O0 -g`) and one with optimizations (using either `-O2` or `-O3`)?"

ChinYikMing commented:
> I am concerned that the proposed changes increase the CI pipeline completion time. Could we rework this for two types of builds: one without optimizations (using `-O0 -g`) and one with optimizations (using either `-O2` or `-O3`)?"

![action-build-time](https://github.com/user-attachments/assets/3853e60c-7e73-438c-8100-9ca44a7fd30a)

When reviewing the build times (see the figure), each takes less than 10 seconds to complete, making it reasonable to include all types. Furthermore, the sum of these additional build time is significantly shorter than the running time on the Arm64 host.

jserv commented:
> When reviewing the build times (see the figure), each takes less than 10 seconds to complete, making it reasonable to include all types. Furthermore, the sum of these additional build time is significantly shorter than the running time on the Arm64 host.

It sounds great for x86-64 host. Let's avoid performing the checks with different optimization levels for Arm64 host though.

ChinYikMing commented:
> > When reviewing the build times (see the figure), each takes less than 10 seconds to complete, making it reasonable to include all types. Furthermore, the sum of these additional build time is significantly shorter than the running time on the Arm64 host.
> 
> It sounds great for x86-64 host. Let's avoid performing the checks with different optimization levels for Arm64 host though.

The current proposed changes only apply to x86-64 host.

jserv commented:
Thank @ChinYikMing for contributing!
